test/functional/teams/segment/tests/test-teams-ListSegment.hpp

```diff
@@ -0,0 +1,143 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-22, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef __TEST_TEAMS_LISTSEGMENT_HPP__
+#define __TEST_TEAMS_LISTSEGMENT_HPP__
+
+/*
+ * NOTE! This functional test is not currently being exercised
+ * see issue https://github.com/LLNL/RAJA/issues/1259 for details
+ */
+
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <ctime>
+#include <algorithm>
+#include <numeric>
+
+template <typename INDEX_TYPE, typename WORKING_RES, typename LAUNCH_POLICY, typename GLOBAL_THREAD_POICY>
+void TeamsListSegmentTestImpl(INDEX_TYPE N)
+{
+
+  // Create and initialize indices in idx_array used to create list segment
+  std::vector<INDEX_TYPE> idx_array;
+
+  srand ( time(NULL) );
+
+  for (INDEX_TYPE i = INDEX_TYPE(0); i < N; ++i) {
+    INDEX_TYPE randval = INDEX_TYPE(rand() % RAJA::stripIndexType(N));
+    if ( i < randval ) {
+      idx_array.push_back(i);
+    }
+  }
+
+  size_t idxlen = idx_array.size();
+
+  camp::resources::Resource working_res{WORKING_RES::get_default()};
+
+  // Create list segment for tests
+  INDEX_TYPE* idx_vals = nullptr;
+  if (N > 0) {
+    idx_vals = &idx_array[0];
+  }
+  RAJA::TypedListSegment<INDEX_TYPE> lseg(idx_vals, idxlen,
+                                          working_res);
+
+  INDEX_TYPE* working_array;
+  INDEX_TYPE* check_array;
+  INDEX_TYPE* test_array;
+
+  size_t data_len = RAJA::stripIndexType(N);
+  if ( data_len == 0 ) {
+    data_len = 1;
+  }
+
+  allocateForallTestData<INDEX_TYPE>(data_len,
+                                     working_res,
+                                     &working_array,
+                                     &check_array,
+                                     &test_array);
+
+  constexpr int threads = 256;
+  int blocks = (data_len - 1)/threads + 1;
+
+  if ( RAJA::stripIndexType(N) > 0 ) {
+
+    for (size_t i = 0; i < idxlen; ++i) {
+      test_array[ RAJA::stripIndexType(idx_vals[i]) ] = idx_vals[i];
+    }
+
+    working_res.memcpy(working_array, test_array, sizeof(INDEX_TYPE) * data_len);
+
+    RAJA::expt::launch<LAUNCH_POLICY>
+      (RAJA::expt::Grid(RAJA::expt::Teams(blocks), RAJA::expt::Threads(threads)),
+        [=] RAJA_HOST_DEVICE(RAJA::expt::LaunchContext ctx) {
+
+        RAJA::expt::loop<GLOBAL_THREAD_POICY>(ctx, lseg, [&](INDEX_TYPE idx) {
```

Comment:
At this point, we're inside a GPU kernel potentially and we are using the list segment in the same way we would pass it from a non-GPU context. The segment indices should be in properly device memory. Do we need to specialize the RAJA::expt::loop construct for list segments?

Reply:
I believe the issue is that we need to capture the list segment into the host_device lambda which triggers the copy and the destructor of the listsegment object. 

https://github.com/LLNL/RAJA/blob/87a5cac67214e5e96c941bd652b1c0981e9f2123/include/RAJA/index/ListSegment.hpp#L174-L180

https://github.com/LLNL/RAJA/blob/87a5cac67214e5e96c941bd652b1c0981e9f2123/include/RAJA/index/ListSegment.hpp#L192-L197

I think what we want is to take a reference of the device pointer and use that within the host_device lambda. 
Perhaps wrapping it into a struct could work?

```
struct list_seg_ptr
{
double * device_ptr; //point to listsegment pointer
int begin;
int end;
};
```

I think we would have the same issue with we tried to use a listsegment inside a raja forall method..
```
RAJA::listSegment lseg;
  RAJA::forall<RAJA::cuda_exec<CUDA_BLOCK_SIZE>>(RAJA::RangeSegment(0, N), 
    [=] RAJA_DEVICE (int i) { 
     lseg.size();  
    c[i] = a[i] + b[i]; 
  });
```    

Let me try the above ^^.

Reply:
Does the `m_owned` tracking not take care of this?

Reply:
I made a small example in PR https://github.com/LLNL/RAJA/pull/1266 , if I use the listsegment directly inside the kernel then I hit this error:
 
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  campCudaErrchk(cudaStreamSynchronize(stream)) unspecified launch failure /g/g17/vargas45/RAJA_DIR/RAJA/tpl/camp/include/camp/resource/cuda.hpp:172
Aborted
```

but if I use the proof concept wrapper it seems to work. @MrBurmark that is a good point about the data not being owned by the object copy, seems like we shouldn't be triggering the free.